### PR TITLE
Initial JNI support

### PR DESCRIPTION
**Pull request by mahesh-hegde**
This PR adds JNI support on Android and Linux.

__Notes on generated files__:
  * The generated files account for majority of code.
  * lib/src/jni_bindings_generated.dart is generated using a personal fork of FFIGEN, which I am not vendoring in this PR because once vendored, we can't see the diff from original. So I have added a git dependency if you want to see the JNI-specific patch to FFIGEN.
  * The bindings can be regenerated by running `dart run ffigen --config ffigen.yaml` in `jni/`. When they are regenerated, there should be no change to any files in the working tree. (Can be verified using git status).
  * It generates bindings for `third_party/jni.h` which is copied from Android NDK and annotated (I copied the file from NDK 23.1, but the file doesn't seem to be updated in a while anyway.
  * I will vendor ffigen_patch_jni in a later PR

  * `lib/src/{jni_object_methods_generated.dart, jni_class_methods_generated.dart, direct_methods_generated.dart}` are generated using a script in tool/ directory. Rerunning it from package root should result in no changes to these files.
 
__Context on platform support__:
  * Android support is the priority of the project. Desktop support is experimental. Currently linux desktop is somewhat well-supported. The package tests pass on linux using the desktop JVM. Dart standalone support is also possible. Example and CI workflow will be added in a later PR.
  * Support for other platforms: tracking issue #10

### CODE REVIEW

**Comment by mahesh-hegde** [comment]
Oops, we have to upload coverage info from both packages. I googled a little. It seems one will be overwritten if we do that.

Can I move the `upload coverage` step to `test_jni` for now? there are no tests in the stub package right now.

**Comment by dcharkes** [comment]
Maybe @liamappelbe as author knows how to deal with multi-packages.

Do we need coverage per subpackage? (With then coverage badges on the nested readme's) Or can we combine it somehow?

**Comment by mahesh-hegde** [comment]
It seems like more of an issue with coveralls. Let me try the parallel build example.

**Comment by mahesh-hegde** [comment]
I'm surprised we even have 25% coverage.

jni_bindings_generated.dart should probably be excluded from coverage because it contains functions for every corner of JNI, many of which are pretty obscure.

I should probably figure out how to generate tests for the generated utility functions using `gen_aux_methods.dart`.

**Comment by dcharkes** [comment]
> Re: badge. The current badge only shows build passing or failing right?

See badge on https://github.com/dart-lang/ffigen

**Comment by mahesh-hegde** [comment]
On surface, coveralls doesn't seem to provide badges for individual jobs.

**Comment by mahesh-hegde** [comment]
There seems to be an issue with windows CI fetching dev dependencies from git. Should not be a problem after ffigen_patch is vendored. Removing that test until then.

File: jni/lib/src/jni_bindings_generated.dart [hunk 1]
```diff
@@ -0,0 +1,3056 @@
...
+// Generated by `package:ffigen`.
+import 'dart:ffi' as ffi;
+
+/// Bindings for libdartjni.so which is part of jni plugin.
+///
+/// It also transitively includes type definitions such as JNIEnv from third_party/jni.h;
+///
+/// However, functions prefixed JNI_ are not usable because they are in a different shared library.
+///
+/// Regenerate bindings with `flutter pub run path_to_ffigen.dart --config ffigen.yaml`.
```
Comment (dcharkes) [inline jni/lib/src/jni_bindings_generated.dart:20]: `flutter pub run ffigen --config ffigen.yaml` and `dart run ffigen --config ffigen.yaml` should work without having to find a path to `ffigen.dart`. (It is the magic of `bin/<package-name>.dart`.)

File: jni/lib/src/jni_object.dart [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  /// [r] still needs to be explicitly deleted when
+  /// it's no longer needed to construct any JniObjects.
+  JniObject.fromGlobalRef(Pointer<JniEnv> env, JniGlobalObjectRef r)
+      : _env = env,
+        _obj = env.NewLocalRef(r._obj),
+        _cls = env.NewLocalRef(r._cls);
+
+  /// Delete the local reference contained by this object.
+  ///
+  /// Do not use a JniObject after calling [delete].
```
Comment (dcharkes) [inline jni/lib/src/jni_object.dart:42]: Add a `boolean _deleted = false;` to `class JniObject`. Set it to true here. And check in all methods that it is false before using it.

Comment (dcharkes) [inline jni/lib/src/jni_object.dart:42]: What happens if I make two `JniObject`s out of one `JObject`, and I call `delete` on one of them? Is that safe? In other words, does the `delete` only delete the `JniObject` object? Or does it also delete the underlying things?

**Review by dcharkes — COMMENTED** [review]
First round of comments.

Super cool that it all works! 🎸 

(JNI is very "reflection-ish", calling method by name and passing an array of arguments. Hopefully JNIgen will provide a fully strongly typed API! 😄)

File: jni/lib/src/jni_object.dart (hunk 2 above)
Comment (mahesh-hegde) [reply jni/lib/src/jni_object.dart:42]: It deletes the underlying reference, i.e original JObject.

I think a warning in documentation is enough. Since there aren't many times you would want to convert a JObject (that's just raw `Pointer<Void>`) -> JniObject. And even then Jni.wrap() is enough.

Safer semantics can be implemented by creating new LocalRef whenever we are copying like this. But that will also prevent us from doing things like `Jni.wrap(low_level_expression_returning_jobject)`, because we will lose original LocalRef and that will leak.

Comment (dcharkes) [reply jni/lib/src/jni_object.dart:42]: Let's wait and see what we generate for end users of JNIgen. If we can hide these implementation details from them, we only have to make sure that we generate the right thing ourselves. If we cannot fully hide this API we should come up with good documentation and a good API to manage these native resources.

Please file a bug for us to look into this later.

File: .github/workflows/test-package.yml [hunk 3]
```diff
@@ -72,27 +72,56 @@ jobs:
...
           channel: 'stable'
+      - uses: actions/setup-java@v2
+        with:
+          distribution: 'zulu'
+          java-version: '11'
       - run: |
           sudo apt-get update -y
           sudo apt-get install -y ninja-build libgtk-3-dev
-      - run: flutter config --enable-linux-desktop
+      - run: |
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply .github/workflows/test-package.yml:101]: Sorry I missed 18 comments because github folded them.

This install and configure step is specific to linux on CI.

File: jni/lib/src/jni.dart [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+/// Load Dart-JNI Helper library.
+///
+/// If path is provided, it's used to load the library.
+/// Else just the platform-specific filename is passed to DynamicLibrary.open
+DynamicLibrary _loadJniHelpersLibrary(
+    {String? dir, String baseName = "dartjni"}) {
+  final fileName = _getLibraryFileName(baseName);
+  final libPath = (dir != null) ? join(dir, fileName) : fileName;
+  final dylib = DynamicLibrary.open(libPath);
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/jni.dart:37]: Can I use exists()?

File: jni/src/CMakeLists.txt [hunk 5]
```diff
@@ -6,12 +6,26 @@ cmake_minimum_required(VERSION 3.10)
 project(jni_library VERSION 0.0.1 LANGUAGES C)
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/src/CMakeLists.txt:6]: Should this be in bin/ or tools/ ?

File: jni/lib/src/jni.dart [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+  ///
+  /// [helperDir] is path of the directory where the wrapper library is found.
+  /// This parameter needs to be passed manually on __Dart standalone target__,
+  /// since we have no reliable way to bundle it with the package.
+  ///
+  /// [jvmOptions], [ignoreUnrecognized], & [jniVersion] are passed to the JVM.
+  /// Strings in [classPath], if any, are used to construct an additional
+  /// JVM option of the form "-Djava.class.path={paths}".
+  static Jni spawn({
+    String? helperDir,
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/jni.dart:109]: Should I do the same for helperDir in Jni.spawn? or is it better kept as string?

[the remark this answers is not in the document]
Comment (dcharkes) [reply jni/lib/src/jni.dart:109]: I prefer `Uri`s everywhere they are just more flexible and precise than a `String`.

**Comment by mahesh-hegde** [comment]
Noted. Thanks.

        Mahesh

File: jni/lib/src/jni_class.dart [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  /// Use this [JniClass] to execute callback, then delete.
+  ///
+  /// Useful in expression chains.
+  T use<T>(T Function(JniClass) callback) {
+    // TODO: Maybe use a mixin or something?
+    // JniClass and JniObject have some similar functionality.
+    _checkDeleted();
+    var result = callback(this);
```
Comment (liamappelbe) [inline jni/lib/src/jni_class.dart:135]: Might be good to catch any exceptions this throws, call `delete()`, then rethrow the exception?

File: jni/src/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/src/CMakeLists.txt:6]: Currently user can specify an explicit source path package_jni/source with -S option.

Is there a way to get path of source in pub cache? `Platform.scipt` returns a path in `.dart-tool`.

Anything else you recommend? Like cloning repo or downloading sources to .dart-tool?

File: jni/lib/src/extensions.dart [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    if (chars == nullptr) {
+      checkException();
+    }
+    final result = chars.cast<Utf8>().toDartString();
+    ReleaseStringUTFChars(jstring, chars);
+    return result;
+  }
+
+  /// Return a new [JString] from contents of [s].
+  JString asJString(String s, [Allocator allocator = malloc]) {
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/extensions.dart:37]: Wow that was an obvious oversight, (allocating with custom allocator but freeing with malloc), thanks.

[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/extensions.dart:37]: Btw, with an Arena don't we want to not actually free memory at the end of scope? Since arena is something passed & managed from outer function.

File: jni/lib/src/jni.dart [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+
+  /// Returns class for [qualifiedName] found by platform-specific mechanism.
+  ///
+  /// TODO: Determine when to use class loader, and when FindClass
+  /// TODO: This isn't consistent with other methods, eg: getCurrentActivity,
+  /// which do not wrap reference in JniObject. But the way it's used most of
+  /// the time, we need a direct method.
+  JniClass findJniClass(String qualifiedName) {
+    var nameChars = qualifiedName.toNativeChars();
+    final cls = _bindings.LoadClass(nameChars);
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/jni.dart:225]: findClass returns a raw JClass (seemed consistent with getApplicationContext etc..), findJniClass returns a JniClass wrapper.

File: jni/lib/src/jni_class.dart [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+import 'jni_object.dart';
+
+part 'jni_class_methods_generated.dart';
+
+final ctorLookupChars = "<init>".toNativeChars();
+
+/// Convenience wrapper around a JNI local class reference.
+///
+/// Reference lifetime semantics are same as [JniObject].
+class JniClass {
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/jni_class.dart:18]: What's the overhead of NativeFinalizers? Are they intended to be used for file descriptor's etc.. resources or more purposes?

I will look into NativeFinalizers.

File: jni/lib/src/jni_class.dart [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+  bool _deleted = false;
+  JniClass.of(this._env, this._cls);
+
+  JniClass.fromJClass(Pointer<JniEnv> env, JClass cls)
+      : _env = env,
+        _cls = cls;
+
+  JniClass.fromGlobalRef(Pointer<JniEnv> env, JniGlobalClassRef r)
+      : _env = env,
+        _cls = env.NewLocalRef(r._cls) {
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/jni_class.dart:30]: This is actually what Daco & I were discussing. There are quite a few edge cases like this, and it makes sense to use Global references entirely, and also provide a thread-safe indirection over JniEnv if possible.

> My understanding is that local refs only last as long as the current native scope. Maybe this doesn't matter too much since the outermost scope will be Dart. But I think it might matter in cases where the user passes a Dart callback to Java, and Java calls back into Dart?

I don't know how to verify the following
>> One unusual case deserves separate mention. If you attach a native thread with AttachCurrentThread, the code you are running will never automatically free local references until the thread detaches. Any local references you create will have to be deleted manually. In general, any native code that creates local references in a loop probably needs to do some manual deletion.

https://developer.android.com/training/articles/perf-jni

But the concern is valid. On Android, thread is attached manually so far in my experimentation, but there may be cases it isn't.

File: jni/lib/src/jni.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/jni.dart:41]: I changed this to catch `Error` on DynamicLibrary.open and rethrow as HelpersNotFoundException. Seemed the best way. dlopen has many cases (eg: specific variables like LD_LIBRARY_PATH and linking flags like RPATH).

File: jni/lib/src/jni_object.dart [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+      _cls = _env.GetObjectClass(_obj);
+    }
+    return _cls;
+  }
+
+  /// Get a JniClass of this object's class.
+  JniClass getClass() {
+    _checkDeleted();
+    if (_cls == nullptr) {
+      return JniClass.of(_env, _env.GetObjectClass(_obj));
```
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/jni_object.dart:86]: I am not saving it in original object, so I think this would be fine.

I need to do NewLocalRef in second case, because `JniObject._cls` also gets deleted (`DeleteLocalRef`) on `JniObject.delete`.

File: jni/lib/src/jni_class.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (dcharkes) [reply jni/lib/src/jni_class.dart:18]: I have been thinking the same thing, but it'd be fine a future thing. Also depends a bit on what the generator will output. E.g. how much do users need to do. So maybe we should postpone making it a `Finalizable` with a `NativeFinalizer` till we can see what kind of code we will generate.

File: jni/lib/src/jni.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (liamappelbe) [reply jni/lib/src/jni.dart:224]: Yeah, but it looks like this function shares a lot of code with `findClass`, so you can do this:

```dart
JniClass findJniClass(String qualifiedName) {
  return JniClass.of(getEnv(), findClass(qualifiedName));
}
```

And `newInstance` can similarly call `findJniClass`

File: jni/lib/src/jni_class.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (liamappelbe) [reply jni/lib/src/jni_class.dart:18]: Agreed, this should be follow up work. Just file a bug to remind yourself to look into it later.

**Review by liamappelbe — COMMENTED** [review]
I think this PR can be submitted soon. You don't have to fix all the things we've talked about before submitting. Big PRs get hard to follow if we try to fix too many things at once. Just make sure you file bugs for anything you're going to do later, so we don't forget it.

As long as the test coverage stuff is working, we can fix the rest bit by bit. The coverage number is artificially low right now, because it's including all the `*_generated.dart` files. Try adding `// coverage:ignore-file` to all those files (put it in the `preamble:` of whatever ffigen config made these files).

**Comment by dcharkes** [comment]
> I think this PR can be submitted soon. You don't have to fix all the things we've talked about before submitting. Big PRs get hard to follow if we try to fix too many things at once. Just make sure you file bugs for anything you're going to do later, so we don't forget it.

Agreed. 🚀 

I'll be out next week, feel free to submit this PR and follow up PRs without me looking at it.

File: jni/lib/src/jni.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (mahesh-hegde) [reply jni/lib/src/jni.dart:224]: Ah didn't notice this. thanks.

**Comment by mahesh-hegde** [comment]
```
Error: Bad response: 504 <html>
<head><title>504 Gateway Time-out</title></head>
<body>
<center><h1>504 Gateway Time-out</h1></center>
</body>
</html>
```

**Comment by dcharkes** [comment]
@mahesh-hegde & @liamappelbe is this ready to be merged? 🎸
